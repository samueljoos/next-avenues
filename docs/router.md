# next-avenues *0.0.1*

> Sofisticated routes for nextjs


### src/Router.js


#### new Router() 

Router is the public interface used to define
routes, groups and Link components.






##### Returns


- `Void`



#### Router._initialize(callback)  *private method*

Initialize the client url location data




##### Parameters

| Name | Type | Description |  |
| ---- | ---- | ----------- | -------- |
| callback | `Function`  |  | &nbsp; |




##### Returns


-  



#### Router._validateGroupClosure(callback)  *private method*

Validates the group closure to make sure
it is a function




##### Parameters

| Name | Type | Description |  |
| ---- | ---- | ----------- | -------- |
| callback | `Function`  |  | &nbsp; |




##### Returns


-  



#### Router._validateNestedGroups()  *private method*

Validates that nested groups are not created.






##### Returns


-  



#### Router.add(route, page) 

Creates a new route which resolves to a Next.js page component.




##### Parameters

| Name | Type | Description |  |
| ---- | ---- | ----------- | -------- |
| route | `String`  |  | &nbsp; |
| page | `String`  |  | &nbsp; |




##### Returns


- `Route`  



#### Router.match(url, host) 

Resolves and returns the route that matches given **url** and **host**

**Note:** The first matching route will be used. So make
sure the generic routes are created after the
static routes.




##### Parameters

| Name | Type | Description |  |
| ---- | ---- | ----------- | -------- |
| url | `String`  |  | &nbsp; |
| host | `String`  |  | &nbsp; |




##### Returns


- `Object`  



#### matchingRoute() 

Find the first matching route.






##### Returns


- `Void`



#### if() 

Return null when unable to find a route.






##### Returns


- `Void`



#### group([name], callback) 

Create a group of routes




##### Parameters

| Name | Type | Description |  |
| ---- | ---- | ----------- | -------- |
| name | `string`  | == null] | *Optional* |
| callback | `Function`  |  | &nbsp; |




##### Examples

```javascript
routes.group(() => {
    routes.add('/', 'dashboard')
}).prefix('admin')
```


##### Returns


- `Group`  



#### group() 

Create a new group and pass all the routes
to the group.






##### Returns


- `Void`



#### list() 

Returns an array of all the registered routes






##### Returns


- `Array`  



#### Link(props) 

Link component for react




##### Parameters

| Name | Type | Description |  |
| ---- | ---- | ----------- | -------- |
| props | `object`  |  | &nbsp; |




##### Returns


-  React.Element



#### getCurrentRoute() 

Return the current active route.
This is usualy called inside the getInitialProps a Next.js page component.






##### Returns


- `object`  



#### getRequestHandler(app, customHandler) 

Middleware function for your nextjs server setup.




##### Parameters

| Name | Type | Description |  |
| ---- | ---- | ----------- | -------- |
| app | `Next.App`  |  | &nbsp; |
| customHandler | `Function`  |  | &nbsp; |




##### Returns


- `Void`




*Documentation generated with [doxdox](https://github.com/neogeek/doxdox).*
