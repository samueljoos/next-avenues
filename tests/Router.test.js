import { createElement } from 'react';
import router from '../src/Router';
import renderer from 'react-test-renderer';
const Link = router.Link;

router.domain = 'local.test';
router.port = 3000;

describe('adding routes', () => {
    it('route can be created', () => {
        const startCount = router.list().length;
        router.add('/home/:slug', 'homepage').as('homepage');
        expect(router.list().length - startCount).toBe(1);
    });

    it('domain route can be created', () => {
        const startCount = router.list().length;
        router.add('/domain/:slug', 'domain').as('domain').domain('local.test');
        expect(router.list().length - startCount).toBe(1);
    });

    it('subdomain route can be created', () => {
        const startCount = router.list().length;
        router.add('/subdomain', 'subdomain').as('subdomain').domain(':domainname.test');
        expect(router.list().length - startCount).toBe(1);
    });

    it('prefixed route can be created', () => {
        const startCount = router.list().length;
        router.add('/prefixed', 'prefix').as('prefix').prefix('prefix');
        expect(router.list().length - startCount).toBe(1);
    });

    it('prefixed domain route can be created', () => {
        const startCount = router.list().length;
        router.add('/domain', 'prefix-domain').as('prefix-domain').domain('local.test').prefix('prefix');
        expect(router.list().length - startCount).toBe(1);
    });

    it('group with routes can be created', () => {
        const startCount = router.list().length;

        router.group(() => {
            router.add('/first', 'group-first').as('group-first');
            router.add('/second', 'group-second').as('group-second');
        });

        expect(router.list().length - startCount).toBe(2);
    });

    it('prefix group with routes can be created', () => {
        const startCount = router.list().length;

        router.group(() => {
            router.add('/first', 'group-prefix-first').as('group-prefix-first');
            router.add('/second', 'group-prefix-second').as('group-prefix-second');
        }).prefix('group-prefix');

        expect(router.list().length - startCount).toBe(2);
    });

    it('domain group with routes can be created', () => {
        const startCount = router.list().length;

        router.group(() => {
            router.add('/domain-first', 'group-domain-first').as('group-domain-first');
            router.add('/domain-second', 'group-domain-second').as('group-domain-second');
        }).domain('local.test');

        expect(router.list().length - startCount).toBe(2);
    });

    it('subdomain group with routes can be created', () => {
        const startCount = router.list().length;

        router.group(() => {
            router.add('/subdomain-first', 'group-subdomain-first').as('group-subdomain-first');
            router.add('/subdomain-second', 'group-subdomain-second').as('group-subdomain-second');
        }).domain(':domainname.test');

        expect(router.list().length - startCount).toBe(2);
    });
});

describe('creating links', () => {
    it('for a route', () => {
        const component = renderer.create(
            createElement(Link, { name: 'homepage', params: { slug: 'slug' }}, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a domain route', () => {
        const component = renderer.create(
            createElement(Link, { name: 'domain', params: { slug: 'slug-2' }}, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a subdomain route', () => {
        const component = renderer.create(
            createElement(Link, { name: 'subdomain', params: { domainname: 'bla' }}, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a prefixed route', () => {
        const component = renderer.create(
            createElement(Link, { name: 'prefix' }, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a prefixed domain route', () => {
        const component = renderer.create(
            createElement(Link, { name: 'prefix-domain' }, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a route in a group', () => {
        const component = renderer.create(
            createElement(Link, { name: 'group-first' }, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a route in a prefix group', () => {
        const component = renderer.create(
            createElement(Link, { name: 'group-prefix-first' }, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a route in a domain group', () => {
        const component = renderer.create(
            createElement(Link, { name: 'group-domain-first' }, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('for a route in a subdomain group', () => {
        const component = renderer.create(
            createElement(Link, { name: 'group-subdomain-first', params: { domainname: 'subdomain' }}, createElement('a', {}, 'link')),
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('resolving routes', () => {
    it('for a route', () => {
        const { page } = router.match('/home/slug', router.domain);
        expect(page).toBe('homepage');
    });

    it('for a domain route', () => {
        const { page } = router.match('/domain/slug-2', router.domain);
        expect(page).toBe('domain');
    });

    it('for a subdomain route', () => {
        const { page } = router.match('/subdomain', 'bla.test');
        expect(page).toBe('subdomain');
    });

    it('for a prefixed route', () => {
        const { page } = router.match('/prefix/prefixed', router.domain);
        expect(page).toBe('prefix');
    });

    it('for a prefixed domain route', () => {
        const { page } = router.match('/prefix/domain', router.domain);
        expect(page).toBe('prefix-domain');
    });

    it('for a route in a group', () => {
        const { page } = router.match('/first', router.domain);
        expect(page).toBe('group-first');
    });

    it('for a route in a prefix group', () => {
        const { page } = router.match('/group-prefix/first', router.domain);
        expect(page).toBe('group-prefix-first');
    });

    it('for a route in a domain group', () => {
        const { page } = router.match('/domain-first', router.domain);
        expect(page).toBe('group-domain-first');
    });

    it('for a route in a subdomain group', () => {
        const { page } = router.match('/subdomain-first', 'subdomain.test');
        expect(page).toBe('group-subdomain-first');
    });
});
