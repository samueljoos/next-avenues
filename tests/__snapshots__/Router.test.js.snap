// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creating links for a domain route 1`] = `
<a
  href="/domain/slug-2"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a prefixed domain route 1`] = `
<a
  href="/prefix/domain"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a prefixed route 1`] = `
<a
  href="http://undefined:3000/prefix/prefixed"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a route 1`] = `
<a
  href="http://undefined:3000/home/slug"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a route in a domain group 1`] = `
<a
  href="/domain-first"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a route in a group 1`] = `
<a
  href="http://undefined:3000/first"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a route in a prefix group 1`] = `
<a
  href="http://undefined:3000/group-prefix/first"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a route in a subdomain group 1`] = `
<a
  href="http://subdomain.test:3000/subdomain-first"
  onClick={[Function]}
>
  link
</a>
`;

exports[`creating links for a subdomain route 1`] = `
<a
  href="http://bla.test:3000/subdomain"
  onClick={[Function]}
>
  link
</a>
`;
